webapi/src/main/scala/org/knora/webapi/slice/admin/domain/service/UserService.scala

```diff
@@ -44,17 +45,32 @@ case class UserService(
   private val userRepo: KnoraUserRepo,
   private val projectsService: ProjectADMService,
   private val groupsService: GroupsResponderADM,
-  private val permissionService: PermissionsResponderADM
+  private val permissionService: PermissionsResponderADM,
+  private val cacheService: CacheService
 ) {
 
   def findUserByIri(iri: UserIri): Task[Option[User]] =
-    userRepo.findById(iri).flatMap(ZIO.foreach(_)(toUser))
+    fromCacheOrRepo(iri, cacheService.getUserByIri, userRepo.findById)
 
   def findUserByEmail(email: Email): Task[Option[User]] =
-    userRepo.findByEmail(email).flatMap(ZIO.foreach(_)(toUser))
+    fromCacheOrRepo(email, cacheService.getUserByEmail, userRepo.findByEmail)
 
   def findUserByUsername(username: Username): Task[Option[User]] =
-    userRepo.findByUsername(username).flatMap(ZIO.foreach(_)(toUser))
+    fromCacheOrRepo(username, cacheService.getUserByUsername, userRepo.findByUsername)
+
+  private def fromCacheOrRepo[A](
+    id: A,
+    fromCache: A => Task[Option[User]],
+    fromRepo: A => Task[Option[KnoraUser]]
+  ): Task[Option[User]] =
+    fromCache(id).flatMap {
+      case Some(user) => ZIO.some(user)
+      case None =>
+        fromRepo(id).flatMap(ZIO.foreach(_)(toUser)).tap {
+          case Some(user) => cacheService.putUser(user)
+          case None       => ZIO.unit
+        }
+    }
```

Comment:
Perhaps?

```
    fromCache(id).flatMap {
      _.map(ZIO.some(_)).getOrElse {
        fromRepo(id).flatMap {
          ZIO.foreach(_)(toUser(_).tap(cacheService.putUser(_)))
        }
      }
    }
```

Reply:
we recently discussed that ZIO code often looks more idiomatic if optionality is moved to the `E` channel. Maybe that could make it even nicer?

Reply:
tbh, I find this less readable.

How about:

```scala
  fromCache(id).flatMap {
      case Some(user) => ZIO.some(user)
      case None       => fromRepo(id).some.flatMap(toUser(_).tap(cacheService.putUser).asSomeError).unsome
    }
    ```